### PR TITLE
Do not call x-focus-frame that confuses the buggy i3 wm (#1641)

### PR DESCRIPTION
**Pull request by errge**
Sorry, but the i3 maintainers decided to play politics and to not even review my changelist from a technical point of view, only from standard committee point of view.  They also told me to get rid of XSetInputFocus from emacs, which is obviously ridiculous, if the rule would be that only WMs are calling that, then the X server would refuse those calls or have an option at least to refuse those calls.

I also understand the issue in #1641 well enough now, to say that it shouldn't happen with most of the other tiling window managers, as i3 contains a bug that is specific to i3.

Fortunately for us, we actually didn't want to call XSetInputFocus, we only called it by accident, so my changelist simply gets rid of that accidental call and fixes the issue that way.

I was able to reproduce the issue with 100% certainty and I can confirm that this change fixes it 100% on my setup.

### CODE REVIEW

**Comment by errge** [comment]
Sorry for the noise, but my fix is not correct yet, I made a mistake, working on it right now...

**Comment by errge** [comment]
Okay, my change is ready to review.

Please note that I had to add a dependency on cl for the flet.  Is this a problem?  If yes, then we can come up with other solutions (e.g. just writing our own select-frame-input-focus without the offending x-focus-frame call).

If flet and cl is not allowed, can you give me some input on why this select-frame-input-focus call is necessary, so I can understand the situation better and come up with a way to fix this?

**Comment by errge** [comment]
Sorry for the noise...

I got rid of the cl dependency, now my changeset only depends on pure emacs lisp and should work in every major emacs version without additional dependencies.

File: helm.el [hunk 1]
```diff
@@ -2449,7 +2449,31 @@ frame configuration as per `helm-save-configuration-functions'."
...
+                 ;; XSetInputFocus call, since we already have focus
+                 ;; for Emacs, the user is just using helm!  We call
+                 ;; select-frame-set-input-focus for the other
+                 ;; side-effects, not for x-focus-frame.
+                 (let ((saved-x-focus-frame (symbol-function 'x-focus-frame)))
+                   (unwind-protect
+                       (progn
+                         (fset 'x-focus-frame #'(lambda (_)))
+                         (select-frame-set-input-focus frame))
+                     (fset 'x-focus-frame saved-x-focus-frame))))))))
```
Comment (thierryvolpiatto) [inline helm.el]: IIUC, you want to disable `x-focus-frame` when calling `select-frame-set-input-focus` ? Right?
If so you can use this instead:

```elisp
(cl-letf (((symbol-function 'x-focus-frame) #'ignore))
  (Do_what_you_have_to_do_here))
```

Comment (thierryvolpiatto) [inline helm.el]: Something like this:

```diff
@@ -2449,3 +2449,4 @@ frame configuration as per `helm-save-configuration-functions'."
-               (let ((frame (if (minibufferp helm-current-buffer)
-                                (selected-frame)
-                              (last-nonminibuffer-frame))))
+               (cl-letf ((frame (if (minibufferp helm-current-buffer)
+                                    (selected-frame)
+                                  (last-nonminibuffer-frame)))
+                         ((symbol-function 'x-focus-frame) #'ignore))
```

**Comment by thierryvolpiatto** [comment]
Thanks to work on this, my knowledge on X server is too limited to comment on `XSetInputFocus` though.
It looks like `x-focus-frame` is not supported by i3 IIUC, probably it would be better to fix this directly in `select-frame-set-input-focus`, but it would take ages and interminable discussion on emacs list, if you are patient enough you can report there.
In the meaning time, let's apply your workaround here in helm.
The new `cl-flet` of cl-lib is much better than `flet` from cl but works only lexically, in our case that's mean that if `x-focus-frame` doesn't appear physically in BODY it will have no effect, in this case the best is to use `cl-letf` (see my comments in diff).

**Comment by errge** [comment]
Hey,

Yes, understanding the XSetInputFocus interaction with all the XCB async usage and X timestamping logic takes a big discussion, which I think should save you, your short summary of "not supported by i3" I think is a good tldr.

You are also right about Emacs, maybe we should propose an extra argument to `select-frame-set-input-focus`, but since it's only 1 wm in the world, I'm not sure how long and time consuming discussion would that be.

Regarding cl-letf:

    (defun x () 42)
    (defun y () (x))
    (flet ((x () 4242)) (y)) ;; 4242, but in 'cl
    (cl-flet ((x () 4242)) (y)) ;; 42, since it's lexical
    (letf (((symbol-function 'x) (lambda () 4242))) (y)) ;; 4242, but in 'cl
    (cl-letf (((symbol-function 'x) (lambda () 4242))) (y)) ;; 4242

Which shows that you are 100% correct, I don't know how I didn't think of it alone.  I had `cl-letf` in my sight, but I failed to combine it with `symbol-function`.  The generated macro code is the same as mine, so thanks for pointing this out, I will fix the pull request and ping you.

Cheers!

**Comment by errge** [comment]
I have updated the pull request.

**Comment by thierryvolpiatto** [comment]
Great!
I have no access to i3 wm, if you think your changes are fixing #1641 feel free to close it as well.
Thanks.

**Comment by errge** [comment]
@thierryvolpiatto Yes, it's fixed, but I don't have a close button on the github ui, maybe only you can close it.

Please close it, i'm pretty sure it's fixed now.

**Comment by thierryvolpiatto** [comment]
Gergely Risko <notifications@github.com> writes:

> @thierryvolpiatto Yes, it's fixed, but I don't have a close button on the github ui, maybe only you can close it.
>
> Please close it, i'm pretty sure it's fixed now.

Ok I am going to close now.

Many thanks for your work!

-- 
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Pull request merged**